hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -2804,9 +2804,15 @@ public MemoryBoundedLogMessageBuffer getRegionServerFatalLogBuffer() {
    * Master runs a coordinated stop of all RegionServers and then itself.
    */
   public void shutdown() throws IOException {
+    if (!isInitialized()) {
+      LOG.info("Shutdown requested but we're not the active master. Proceeding as a stop.");
```

Comment:
I think another reasonable alternative here is to reject the RPC back to the client with some form of `DoNotRetryIOException` and a message about cannot shutdown from an inactive master.

What do you think?

Reply:
Well, I'm not entirely sure if this is the intended way to stop stand-by masters. Reading the code.. it looks like the active master coordinates a proper cluster shutdown. In that process, it removes the /hbase/running znode that the standb-bys keep a watch on. See the following code in ActiveMasterManager..

```
@Override
  public void nodeDeleted(String path) {

    // We need to keep track of the cluster's shutdown status while
    // we wait on the current master. We consider that, if the cluster
    // was already in a "shutdown" state when we started, that this master
    // is part of a new cluster that was started shortly after the old cluster
    // shut down, so that state is now irrelevant. This means that the shutdown
    // state must be set while we wait on the active master in order
    // to shutdown this master. See HBASE-8519.
    if(path.equals(watcher.getZNodePaths().clusterStateZNode) && !master.isStopped()) {
      clusterShutDown.set(true);
    }

```
Ideally they should shut themselves down if the ZK event notifications happen as expected. Is that not the case?

Reply:
Yeah, this stuff is tricky and there be dragons if you disturb the current order.

The standbys don't go down currently?

Reply:
@bharathv has the gist of it. At the point of this race condition -- all four of these fields are `null` at the time the rpc is received -- the master will simply do nothing. However, any master (active or backup) can currently receive the rpc and if it's `clusterStatusTracker` is non-null, it will delete this ZK node. From there, in the case of a backup master, the `ActiveMasterManager` will notice and `stop` itself.

Related, looks like there's an early-out in `ServerManager#shutdown` that can result in a master `stop`ping without properly shutting down its procedure store.

```java
    if (onlineServers.isEmpty()) {
      // we do not synchronize here so this may cause a double stop, but not a big deal
      master.stop("OnlineServer=0 right after cluster shutdown set");
    }
```